it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java

```diff
@@ -0,0 +1,220 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.it.pubsub;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.cloud.pubsub.v1.SubscriptionAdminClient;
+import com.google.cloud.pubsub.v1.SubscriptionAdminSettings;
+import com.google.cloud.pubsub.v1.TopicAdminClient;
+import com.google.cloud.pubsub.v1.TopicAdminSettings;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PushConfig;
+import com.google.pubsub.v1.Subscription;
+import com.google.pubsub.v1.SubscriptionName;
+import com.google.pubsub.v1.Topic;
+import com.google.pubsub.v1.TopicName;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default class for implementation of {@link PubsubResourceManager} interface.
+ *
+ * <p>The class provides an interaction with the real Pub/Sub client, with operations related to
+ * management of topics and subscriptions.
+ */
+public final class DefaultPubsubResourceManager implements PubsubResourceManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPubsubResourceManager.class);
+
+  private static final int DEFAULT_ACK_DEADLINE_SECONDS = 600;
+
+  private final String testId;
+  private final String projectId;
+  private final CredentialsProvider credentialsProvider;
+  private final TopicAdminClient topicAdminClient;
+  private final SubscriptionAdminClient subscriptionAdminClient;
+
+  private final List<TopicName> createdTopics;
+  private final List<SubscriptionName> createdSubscriptions;
+
+  private DefaultPubsubResourceManager(Builder builder) throws IOException {
+    this(builder.testId, builder.projectId, builder.credentialsProvider, null, null);
+  }
+
+  @VisibleForTesting
+  DefaultPubsubResourceManager(
+      @Nullable String testId,
+      String projectId,
+      CredentialsProvider credentialsProvider,
+      @Nullable TopicAdminClient topicAdminClient,
+      @Nullable SubscriptionAdminClient subscriptionAdminClient)
+      throws IOException {
+    this.projectId = projectId;
+    this.testId = (!Strings.isNullOrEmpty(testId) ? testId : "");
```

Comment:
An empty id should be an error, and it's probably best to error closer to where the problem arose:

```
public static Builder builder(String projectId, String testId) {
  checkArgument(!Strings.isNullOrEmpty(testId), ...);
  // Other checks
  return new Builder(projectId, testId);
}
```